dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h

```diff
@@ -0,0 +1,211 @@
+    for (int i = 0; i < shapes.num_samples(); i++) {
+      ret[i] = whole_image(shapes[i]);
+    }
+  } else {
+    for (size_t i = 0; i < rois.size(); i++) {
+      ret[i] = intersection(rois[i], whole_image(shapes[i]));
+    }
+  }
+
+  return ret;
+}
+
+
+template <class OutputType, class InputType, int ndims>
+std::vector<SampleDescriptor<InputType, OutputType, ndims - 1>>
+CreateSampleDescriptors(const InListGPU<InputType, ndims> &in,
+                        const OutListGPU<OutputType, ndims> &out,
+                        const std::vector<float> &brightness, const std::vector<float> &contrast) {
+  std::vector<SampleDescriptor<InputType, OutputType, ndims - 1>> ret(in.num_samples());
+
+  for (int i = 0; i < in.num_samples(); i++) {
+    auto &sample = ret[i];
+    sample.in = in[i].data;
+    sample.out = out[i].data;
+    sample.in_pitch = {};
+    sample.out_pitch = {};
+
+    auto fill_pitch_with_flattening = [](const auto &tv, auto &pitch) {
+        auto tl_end = end(tv.shape);
+        auto pitch_end = pitch.end();
+        std::copy(begin(tv.shape), end(tv.shape), pitch.begin());
+        *--pitch_end *= *--tl_end;
+    };
+
+    fill_pitch_with_flattening(in[i], sample.in_pitch);
+    fill_pitch_with_flattening(out[i], sample.out_pitch);
+
+    sample.brightness = brightness[i];
+    sample.contrast = contrast[i];
+  }
+
+  return ret;
+}
+
+
+template <class InputType, class OutputType, int ndims>
+__global__ void
+BrightnessContrastKernel(const SampleDescriptor<InputType, OutputType, ndims> *samples,
+                         const BlockDesc<ndims> *blocks) {
+  auto block = blocks[blockIdx.x];
+  auto sample = samples[block.sample_idx];
+
+  auto *__restrict__ in = sample.in;
+  auto *__restrict__ out = sample.out;
+
+  for (int y = threadIdx.y + block.start.y; y < threadIdx.y + block.end.y; y += blockDim.y) {
+    for (int x = threadIdx.x + block.start.x; x < threadIdx.x + block.end.x; x += blockDim.x) {
+      out[y * sample.out_pitch.x + x] = ConvertSat<OutputType>(
+              in[y * sample.in_pitch.x + x] * sample.contrast + sample.brightness);
+    }
+  }
+}
+
+
+template <typename InputType, typename OutputType, int ndims>
+class BrightnessContrastGpu {
+ private:
+  static constexpr size_t spatial_dims = ndims - 1;
+  using BlockDesc = kernels::BlockDesc<spatial_dims>;
+
+  std::vector<SampleDescriptor<InputType, OutputType, ndims>> sample_descriptors_;
+
+ public:
+  BlockSetup<spatial_dims, -1 /* No channel dimension, only spatial */> block_setup_;
+
+
+  KernelRequirements Setup(KernelContext &context, const InListGPU<InputType, ndims> &in,
+                           const std::vector<float> &brightness, const std::vector<float> &contrast,
+                           const std::vector<Roi<spatial_dims>> &rois = {}) {
+    DALI_ENFORCE(rois.empty() || rois.size() == static_cast<size_t>(in.num_samples()),
+                 "Provide ROIs either for all or none input tensors");
+    DALI_ENFORCE([=]() -> bool {
```

Comment:
1. I find a better practice to manually capture what you need when using lambdas, `[roi]` or `[&roi]`
2. I think it would be better if you rephrase this to give a more specialized message :
```
for (const auto &roi, rois) {
    DALI_ENFORCE(all_coord(roi.hi >= roi.lo)), "Invalid region of interest: {print the values here}"); 
}
```
That should give you a more useful message

Reply:
Will fix in next PR